switchbot/__init__.py

```diff
@@ -220,42 +313,47 @@ def press(self) -> bool:
         return self._sendcommand(PRESS_KEY, self._retry_count)
 
     def switch_mode(self) -> str:
-        """Return Toggle or Switch from cache.
-        Run update first."""
-        return self._mode
+        """Return true or false from cache."""
+        # To get actual position call update() first.
+        if not self._switchbot_device_data:
+            self.update()
+        return self._switchbot_device_data["data"]["switchMode"]
 
     def is_on(self) -> bool:
-        """Return switch state from cache.
-        Run update first."""
-        return self._is_on
+        """Return switch state from cache."""
```

Comment:
I would not expect `is_on` to connect to the device, but return the latest state

Reply:
https://github.com/Danielhiversen/pySwitchbot/blob/9044c5c86d27e56801ba2a085ac9c10315824eca/switchbot/__init__.py#L190

Scanning for advertisement data returns all devices with one scan. The "discover" method returns data for all devices in one 5 second scan.  (Why do multiple scans when one is all that is needed?)

https://github.com/Danielhiversen/pySwitchbot/blob/17f053fcdfce4d9ff88a173e13c92edef31c633c/switchbot/__init__.py#L184
switchbot.SwitchbotDevice().discover() -> returns data for all devices and stores it in self._all_services_data = {}

https://github.com/Danielhiversen/pySwitchbot/blob/17f053fcdfce4d9ff88a173e13c92edef31c633c/switchbot/__init__.py#L276
The is_on fuction returns self._switchbot_device_data["data"]["isOn"] from cache.

Apologies for the long explanation.